src/Libraries/DSCPython/CPythonEvaluator.cs

```diff
@@ -10,6 +10,119 @@
+    [IsVisibleInDynamoLibrary(false)]
+    public class DynamoCPythonHandle : IDisposable
+    {
+        /// <summary>
+        /// A static map of DynamoCPythonHandle counts, is used to avoid removing the underlying python objects from the 
+        /// global scope while there are still handles referencing them.
+        /// </summary>
+        internal static Dictionary<DynamoCPythonHandle, int> HandleCountMap = new Dictionary<DynamoCPythonHandle, int>(new DynamoCPythonHandleComparer());
+
+        /// <summary>
+        /// A unique ID that identifies this python object. It's as a lookup
+        /// symbol within the global scope to find this python object instance.
+        /// </summary>
+        internal IntPtr PythonObjectID { get; set; }
+        /// <summary>
+        /// A string representation of this object.
+        /// TODO - it may be better to just call ToString - this was an attempt
+        /// to optimize needing to use locks and the GIL to get strings for preview bubbles.
+        /// </summary>
+        internal string StringRepresentation { get; set; }
+        public DynamoCPythonHandle(IntPtr id,string strRep)
+        {
+            PythonObjectID = id;
+            StringRepresentation = strRep;
+            if (HandleCountMap.ContainsKey(this)) {
+                HandleCountMap[this] = HandleCountMap[this] +1;
+            }
+            else
+            {
+                HandleCountMap.Add(this, 1);
+            }
+           
+        }
+
+        public override string ToString()
+        {
+            if (String.IsNullOrEmpty(StringRepresentation))
+            {
+                return $"CPython Object ID:{PythonObjectID.ToString()}";
+            }
+            else
+            {
+                return StringRepresentation;
+            }
+        }
+
+        /// <summary>
+        /// When this handle goes out of scope
+        /// we should remove the pythonObject from the globalScope.
+        /// In most cases the DSVM will call this.
+        /// </summary>
+        public void Dispose()
+        {
+            //key doesn't exist.
+            if (!HandleCountMap.ContainsKey(this)){
+                return;
+            }
+            //there are more than 1 reference left, don't dispose.
+            //decrement refs
+            if(HandleCountMap[this] > 1)
+            {
+                HandleCountMap[this] = HandleCountMap[this] - 1;
+                return;
+            }
+
+            IntPtr gs = PythonEngine.AcquireLock();
+            try
+            {
+                using (Py.GIL())
+                {
+
+                    PyScopeManager.Global.Get(CPythonEvaluator.globalScopeName).Remove(PythonObjectID.ToString());
+                    HandleCountMap.Remove(this);
+                   
+                } 
+            }
+            catch (Exception E)
+            {
+                System.Diagnostics.Debug.WriteLine("error removing python object from global scope");   
```

Comment:
hmm @mmisol - because this mostly happens when the graph changes or when nodes are deleted - putting a warning on the node is basically useless -  the only thing I can think of here at the moment is to implement the `ILogSource` interface or pass a logger to the CPythonEvaluator class, that will take a bit of refactoring and the DSCPython binary will then reference DynamCore.

Any other suggestions?

Reply:
You are right, I forgot about that. I guess we can leave that for the future. Maybe log a ticket?

Reply:
I'll add a TODO and a ticket number.